### PR TITLE
[#348] Data Ingestion Redesign

### LINKED ISSUE #348: Data Ingestion redesign
Rethink and rework all non-fixed-sized parts of recommender like itemspace in context of events and NNs inputs.
-  Redesign users and services features engineering part.
-  Use NLP.
-  Untangle vector shapes from db schema and ids.

### PR DESCRIPTION
**Pull request by JanKapala**
Closes #348

### CODE REVIEW

File: recommender/pipeline_configs.py [hunk 1]
```diff
@@ -126,33 +104,10 @@
     REWARD_LOWER_BOUND,
 )
 
-SE = 128
-UE = 64
+SE = Services2tensorsEmbedder().embedding_dim
```
Comment (wujuu) [inline recommender/pipeline_configs.py:107]: Shouldnt this be the other way around? Embedders getting their dimensions from config I mean

**Review by wujuu — CHANGES_REQUESTED** [review]
Overall LGTM, however could we fix the static 512 embedding dim first, before we merge? Also don't forget about a secondary main branch for the continous learning support feature

File: recommender/engines/nlp_embedders/embedders.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        index_id_map = create_index_id_map(objects)
+        [handler(obj) for obj in objects for handler in self.handlers]
+        objects_tensors = sum([handler.get_results() for handler in self.handlers], [])
+        if concat:
+            objects_tensor = torch.cat(objects_tensors, 1)
+            # return objects_tensor, index_id_map
+            return (
+                objects_tensor[:, :512],
+                index_id_map,
+            )  # TODO: this is hard cut, only for smaller size, it should be
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/embedders.py:57]: This code need more profound changes so I will implement them in the next NLP-related PR. This hardcut is just a workaround and I will not introduce proposed arguments because it would be extending the existing workaround rather than solving problem. Such a workaround would be pruned in the next PR anyway.

[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/embedders.py:57]: Btw I will keep these TODOs to remember about it in the next PR

File: recommender/engines/nlp_embedders/handlers.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+
+class ListOfStringsHandler(Handler):
+    def __init__(self, field_name, text2vec=None):
+        super().__init__(text2vec)
+        self.field_name = field_name
+        self.accumulator = []
+
+    def __call__(self, obj):
+        self.accumulator.append(self.text2vec(", ".join(getattr(obj, self.field_name))))
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/handlers.py:70]: Yes, it probably shouldn't be treated in this way - this is the scope of the next NLP-related PR

File: recommender/engines/nlp_embedders/handlers.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    def get_results(self) -> List[Tensor]:
+        result = [torch.stack(self.accumulator)]
+        self.accumulator = []
+        return result
+
+
+class ObjectHandler(Handler):
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/handlers.py:78]: No and for now it should be done in this way because 1) it is possible for now (no information loss) 2) it is very beneficial because it makes possible online learning from unseen examples

File: recommender/engines/nlp_embedders/handlers.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            if subobjects:
+                seq = []
+                for subobj in subobjects:
+                    subfield_val = getattr(subobj, subfield_name)
+                    if subfield_val is None:
+                        subfield_val = ""
+                    seq.append(self.text2vec(subfield_val))
+                tensor = torch.stack(seq)
+            else:
+                tensor = torch.zeros(1, self.MAGIC_NUMBER)
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/handlers.py:134]: This is part of the same story as in the case of hard-cut workaround (explained in one of the previous comments and answers) - it will be refactored in the next NLP-related PR

File: recommender/engines/nlp_embedders/text2vec.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    def __init__(self):
+        self.nlp = spacy_universal_sentence_encoder.load_model("en_use_lg")
+
+    def __call__(self, text: str) -> Tensor:
+        return Tensor(self.nlp(text).vector)
+
+    @property
+    def embedding_dim(self):
+        return 512  # TODO: come on... did you think about doing it correctly?
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/text2vec.py:22]: This is part of the same story as in the case of hard-cut workaround (explained in one of the previous comments and answers) - it will be refactored in the next NLP-related PR

File: recommender/engines/rl/ml_components/synthetic_dataset/dataset.py [hunk 7]
```diff
@@ -173,13 +170,3 @@ def generate_synthetic_sarses(
             state = next_state
 
     return sarses
-
-
-def _embed_and_normalize(service_embedder):
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/rl/ml_components/synthetic_dataset/dataset.py:178]: yes

File: recommender/engines/rl/training/model_evaluation_step/model_evaluation_step.py [hunk 8]
```diff
@@ -123,7 +100,7 @@ def _evaluate_recommendation_time(self, actor, sarses):
         for _ in range(self.time_measurement_samples):
             start = time()
             example_user = random.choice(sarses).state.user
-            elastic_services = [service.id for service in Service.objects]
+            elastic_services = tuple(int(service.id) for service in Service.objects)  # TODO: so.. it works?
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/rl/training/model_evaluation_step/model_evaluation_step.py:103]: yes it is :)

File: recommender/pipeline_configs.py (hunk 1 above)
Comment (JanKapala) [reply recommender/pipeline_configs.py:107]: Yes and no. Embedders should get  their dimension from config but on the other hand `Services2tensorsEmbedder` class uses NLP engine and language corpus that has some specific dimension of output tensors so it can't be set. Maybe in the next NLP-related PR it could be solved in some way but on the other hand It is not necessary needed.

File: recommender/engines/nlp_embedders/embedders.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    ListOfStringsHandler,
+    ObjectHandler,
+)
+from recommender.engines.rl.utils import create_index_id_map
+
+SERVICE_EMBEDDING_DIM = "SE"
+USER_EMBEDDING_DIM = "UE"
+
+
+class Objects2tensorsEmbedder:
```
[the remark this answers is not in the document]
Comment (JanKapala) [reply recommender/engines/nlp_embedders/embedders.py:25]: Not necessarily, there is a bunch of possible handlers (5) and it would explode to 5 classes + abstract one while it is perfectly fine to inject handlers as a strategy and keep classes hierarchy as simple as possible.
